crates/sui-core/src/consensus_adapter.rs

```diff
@@ -226,48 +256,82 @@ impl ConsensusAdapter {
+    /// The function targets having 1 or 2 consensus transaction submitted per user transaction
     /// when system operates normally.
     ///
-    /// The function returns the delay to wait and the position of authority in the list.
-    fn submit_delay_certificate(
+    /// The function returns the position of this authority when it is their turn to submit the transaction to consensus.
+    fn submission_position(
+        &self,
         committee: &Committee,
         ourselves: &AuthorityName,
         tx_digest: &TransactionDigest,
-    ) -> (Duration, usize) {
-        let position = position_submit_certificate(committee, ourselves, tx_digest);
-        const MAX_DELAY_MUL: usize = 10;
-        // DELAY_STEP is chosen as 1.5 * mean consensus delay
-        // In the future we can actually use information about consensus rounds instead of this delay
-        const DELAY_STEP: Duration = Duration::from_secs(7);
-        (
-            DELAY_STEP * std::cmp::min(position, MAX_DELAY_MUL) as u32,
-            position,
-        )
+    ) -> usize {
+        let positions = order_validators_for_submission(committee, tx_digest);
+
+        self.check_submission_wrt_connectivity(ourselves, positions)
+    }
+
+    /// This function runs the following algorithm to decide whether or not to submit a transaction
+    /// to consensus.
+    ///
+    /// It takes in a deterministic list that represents positions of all the authorities.
+    /// The authority in the first position will be responsible for submitting to consensus, and
+    /// so we check if we are this validator, and if so, return true.
+    ///
+    /// If we are not in that position, we check our connectivity to the authority in that position.
+    /// If we are connected to them, we can assume that they are operational and will submit the transaction.
+    /// If we are not connected to them, we assume that they are not operational and we will not rely
+    /// on that authority to submit the transaction. So we shift them out of the first position, and
+    /// run this algorithm again on the new set of positions.
+    ///
+    /// This can possibly result in a transaction being submitted twice if an authority sees a false
+    /// negative in connectivity to another, such as in the case of a network partition.
+    ///
+    /// Recursively, if the authority further ahead of us in the positions is a low performing authority, we will
+    /// move our positions up one, and submit at the same time. This allows low performing
+    /// node a chance to participate in consensus and redeem their scores while maintaining performance.
+    fn check_submission_wrt_connectivity(
+        &self,
+        ourselves: &AuthorityName,
+        positions: Vec<AuthorityName>,
+    ) -> usize {
+        let filtered_positions = positions.into_iter().filter(|authority| {
+            self.connection_monitor_status
```

Comment:
I think it would be more obvious if we just do check for `ourselves` in this code block, instead of passing it to the `check_connection`.
E.g. `ourselves` can be removed from `check_connection` and this line can be rewritten as following so that it is perfectly clear the `expect` below does not panic:

```let filtered_positions = positions.into_iter().filter(|authority| {
            ourselves == authority || // Always include ourselves
            self.connection_monitor_status
                .check_connection(authority) == Some(ConnectionStatus::Connected)
});
```

Reply:
ok, I am adding this but also keeping the other check for safety